sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -1430,7 +1430,18 @@ class Dataset[T] private[sql](
    */
   @scala.annotation.varargs
   def select(cols: Column*): DataFrame = withPlan {
-    Project(cols.map(_.named), logicalPlan)
+    val untypedCols = cols.map {
+      case typedCol: TypedColumn[_, _] =>
+        if (!typedCol.needInputType) {
```

Comment:
just noticed: why don't we inline this method? Then we can centralize the changes here. The methods in `TypedColumn` can still be accessed by java users who ignore "private[spark]"

Reply:
Oh ok. I previously don't want to make `select` looks complicated. Inlined it now.